scrappy/scrappy.pyx

```diff
@@ -0,0 +1,740 @@
+                    # of the image
+                    if numpix == 0:
+                        cleanarr[j, i] = backgroundlevel
+                    else:
+                        # else take the mean
+                        cleanarr[j, i] =  cymedian(medarr, numpix)
+                        #cleanarr[j, i] = PyMedian(medarr, numpix)
+        free(medarr)
+
+
+cdef void clean_idwinterp(float[:, ::1] cleanarr, bool[:, ::1] crmask,
+                          bool[:, ::1] mask, int nx, int ny,
+                          float backgroundlevel):
+    """clean_idwinterp(cleanarr, crmask, mask, nx, ny, backgroundlevel)\n
+    Clean the bad pixels in cleanarr using a 5x5 using inverse distance
+    weighted interpolation.
+
+    Parameters
+    ----------
+    cleanarr : float numpy array
+        The array to be cleaned.
+
+    crmask : boolean numpy array
+        Cosmic ray mask. Pixels with a value of True in this mask will be
+        cleaned.
+
+    mask : boolean numpy array
+        Bad pixel mask. Values of True indicate bad pixels.
+
+    nx : int
+        Size of cleanarr in the x-direction (int). Note cleanarr has dimensions
+        ny x nx.
+
+    ny : int
+        Size of cleanarr in the y-direction (int). Note cleanarr has dimensions
+        ny x nx.
+
+    backgroundlevel : float
+        Average value of the background. This value will be used if there are
+        no good pixels in a 5x5 region.
+    """
+
+    # Go through all of the pixels, ignore the borders
+    cdef int i, j, k, l
+    cdef float f11, f12, f21, f22 = backgroundlevel
+    cdef int x1, x2, y1, y2
+    weightsarr = np.array([[0.35355339, 0.4472136, 0.5, 0.4472136, 0.35355339],
+                          [0.4472136, 0.70710678, 1., 0.70710678, 0.4472136],
+                          [0.5, 1., 0., 1., 0.5],
+                          [0.4472136, 0.70710678, 1., 0.70710678, 0.4472136],
+                          [0.35355339, 0.4472136, 0.5, 0.4472136, 0.35355339]],
+                          dtype=np.float32)
+    cdef float[:, ::1] weights = weightsarr
+    cdef float wsum
+    cdef float val
+    cdef int x, y
+    # For each pixel
+    with nogil, parallel():
+
+        for j in prange(2, ny - 2):
+            for i in range(2, nx - 2):
+                # if the pixel is in the crmask
+                if crmask[j, i]:
+                    wsum = 0.0
+                    val = 0.0
+                    for l in range(-2, 3):
+                        y = j + l
+                        for k in range(-2, 3):
+                            x = i + k
+                            if not (crmask[y, x] or mask[y, x]):
+                                val = val + weights[l, k] * cleanarr[y, x]
+                                wsum = wsum + weights[l, k]
+                    if wsum < 1e-6:
+                        cleanarr[j, i] = backgroundlevel
+                    else:
+                        cleanarr[j, i] = val / wsum
+
+
+def gausskernel(float psffwhm, int kernsize):
```

Comment:
That is a fair point. They don't really need to be ported over to Cython as these calculations are small and only need to be done once. One concern I have is time scale. If I do a pull request to add Moffat to the astropy.convolution when would that get released? Also, some of the utils functions could be quite useful in the core, like the median filters etc (though they would need some modification). At one point we discussed leaving the duplication for now, but beginning to port over the functionality over time. Discussion?